src/Orchard.Web/Core/Contents/Controllers/AdminController.cs

```diff
@@ -13,20 +13,19 @@
 using Orchard.Core.Contents.Settings;
 using Orchard.Core.Contents.ViewModels;
 using Orchard.Data;
-using Orchard.DisplayManagement;
 using Orchard.Localization;
+using Orchard.Localization.Services;
 using Orchard.Logging;
 using Orchard.Mvc.Extensions;
 using Orchard.Mvc.Html;
+using Orchard.Settings;
 using Orchard.UI.Navigation;
 using Orchard.UI.Notify;
-using Orchard.Settings;
 using Orchard.Utility.Extensions;
-using Orchard.Localization.Services;
 
 namespace Orchard.Core.Contents.Controllers {
     [ValidateInput(false)]
-    public class AdminController : Controller, IUpdateModel {
+    public class AdminController : ContentControllerBase, IUpdateModel {
```

Comment:
I hate inherited controllers. I much prefer reusing the code.

Reply:
What's wrong with it? Reusing the code is exactly the point of inheritance in this case and I would consider this logic specific to this feature (these two controllers).

Like I said above, this could also be in a helper class, but all other controllers in Orchard (e.g. the ones in Orchard.Blogs) are specific to a limited set of content types (usually one), so they wouldn't need to redirect when called with an Id that doesn't match the content type they are responsible for, just return a 404. And I would consider this also true for "3rd-party" controllers too, at least I haven't seen (or written) any that would try to redirect instead of 404.